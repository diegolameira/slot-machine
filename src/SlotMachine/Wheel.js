import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

class Wheel extends PureComponent {
  static propTypes = {
    className: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  };

  render() {
    const {
      props: { className, items },
    } = this;
    return (
      <div className={className}>
        {items.map(item => (
          <WheelItem key={item} img={item} />
        ))}
      </div>
    );
  }
}

export const WheelItem = styled.img.attrs(({ img }) => ({
  src: `/imgs/${img}.jpg`,
}))`
  position: absolute;
  background: green;
  width: 80%;
  height: 80%;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;

export default styled(Wheel)`
  position: relative;
  overflow: hidden;
  width: 100%;
  height: 0;
  padding-bottom: 33%;
`;
