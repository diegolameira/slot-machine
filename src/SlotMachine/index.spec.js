import React from 'react';
import TestRenderer from 'react-test-renderer';
import diff from 'lodash.difference';
import shuffle from 'lodash.shuffle';

import { SlotMachine } from './index';
import Wheel, { WheelItem } from './Wheel';

const items = ['strawberry', 'orange', 'banana', 'monkey'];
const buildWheels = (length = 3) =>
  Array.from({ length }, () => shuffle(items));

describe('SlotMachine', () => {
  const start = () => () => false;
  const stop = () => () => false;
  const wheels = buildWheels(3);

  const component = TestRenderer.create(
    <SlotMachine {...{ wheels, stop, start }} />
  );

  it('should render properly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should have 3 (or given) wheels', () => {
    let wheels = component.root.findAllByType(Wheel);
    expect(wheels.length).toBe(3);
    component.update(<SlotMachine wheels={buildWheels(2)} />);
    wheels = component.root.findAllByType(Wheel);
    expect(wheels.length).toBe(2);
  });

  it('should have a start button', () => {});
  it('should have a stop button', () => {});

  describe('Wheels', () => {
    it('should have 4 symbols each (strawberry, banana, orange and a monkey)', () => {
      const items = ['strawberry', 'banana', 'orange', 'monkey'];

      component.root.findAllByType(Wheel).forEach(w => {
        let _items = w.findAllByType(WheelItem);
        expect(_items.length).toBe(items.length);
        let itemsImage = [];
        _items.map(i => itemsImage.push(i.props.img));
        expect(itemsImage.length).toBe(items.length);
        expect(diff(items, itemsImage).length).toBe(0);
      });
    });
    it('starts with the symbols in some random position', () => {
      const wheelsOrder = [];

      component.root.findAllByType(Wheel).forEach(w => {
        let items = w.findAllByType(WheelItem);
        let itemsImage = [];
        items.map(i => itemsImage.push(i.props.img));
        wheelsOrder.push(itemsImage.join(''));
      });

      const areEqual = wheelsOrder.every((val, i, arr) => val === arr[0]);

      expect(areEqual).toBe(false);
    });
  });

  describe('Start button', () => {
    it('should spin the wheels on click (a symbol every, 50ms)', () => {});
    it('should automatically start after 5 seconds', () => {});
  });

  describe('Stop button', () => {
    it('should stop wheels from spinning on click', () => {});
    it('should automatically stop wheels from spinning after 10 seconds (after starting)', () => {});
  });

  describe('Prizes', () => {
    it('10 dollars when two identical non-consecutive symbols', () => {});
    it('20 dollars when two consecutive symbols', () => {});
    it('100 dollars when same symbol in all the wheels', () => {});
  });
});
