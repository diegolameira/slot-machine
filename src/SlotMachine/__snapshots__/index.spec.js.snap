// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotMachine should render properly 1`] = `
<div
  className="sc-htpNat imLdsl"
>
  <header>
    <h1>
      Slot Machine
    </h1>
  </header>
  <main>
    <div
      className="sc-bwzfXH kgSWDu"
    >
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/strawberry.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/orange.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/banana.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/monkey.jpg"
      />
    </div>
    <div
      className="sc-bwzfXH kgSWDu"
    >
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/strawberry.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/orange.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/banana.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/monkey.jpg"
      />
    </div>
    <div
      className="sc-bwzfXH kgSWDu"
    >
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/strawberry.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/orange.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/banana.jpg"
      />
      <img
        className="sc-bdVaJa eqvFIQ"
        src="/imgs/monkey.jpg"
      />
    </div>
  </main>
  <button
    className="sc-bxivhb gxrYMO"
    onClick={[Function]}
  >
    Start
  </button>
  <button
    className="sc-bxivhb gxrYMO"
    onClick={[MockFunction]}
  >
    Stop
  </button>
</div>
`;
